array.c

```diff
@@ -3855,6 +3873,82 @@ rb_ary_values_at(int argc, VALUE *argv, VALUE ary)
     return result;
 }
 
+static VALUE
+append_fetch_values_single(VALUE result, VALUE ary, long olen, VALUE idx, bool block_given)
+{
+    long beg, len;
+    if (FIXNUM_P(idx)) {
+        beg = FIX2LONG(idx);
+    }
+    /* check if idx is Range */
+    else if (rb_range_beg_len(idx, &beg, &len, olen, 1)) {
+        if (NIL_P(RANGE_BEG(idx))) {
+            rb_raise(rb_eIndexError, "Can't use beginless ranges with Array#fetch_values");
+        } else if (NIL_P(RANGE_END(idx))) {
+            rb_raise(rb_eIndexError, "Can't use endless ranges with Array#fetch_values");
+        }
+
+        if (len > 0) {
+            long index;
+            for (index = 0; index < len; index++) {
+                rb_ary_push(result, ary_fetch(ary, beg + index, Qundef, block_given));
+            }
+        }
+        return result;
+    }
+    else {
+        beg = NUM2LONG(idx);
+    }
+    return rb_ary_push(result, ary_fetch(ary, beg, Qundef, block_given));
+}
+
+/*
+ * call-seq:
+ *   fetch_values(*indexes) -> new_array
+ *   fetch_values(*indexes) { |index| ... } -> new_array
+ *
+ * With no block given, returns a new array containing the elements of +self+
+ * at the offsets specified by +indexes+. Each of the +indexes+ must be an
+ * {integer-convertible object}[rdoc-ref:implicit_conversion.rdoc@Integer-Convertible+Objects]:
+ *
+ *    a = [:foo, :bar, :baz]
+ *    a.fetch_values(2, 0)   # => [:baz, :foo]
+ *    a.fetch_values(2.1, 0) # => [:baz, :foo]
+ *    a.fetch_values         # => []
+ *
+ * For a negative index, counts backwards from the end of the array:
+ *
+ *    a.fetch_values(-2, -1) # [:bar, :baz]
+ *
+ * When no block is given, raises an exception if any index is out of range.
+ *
+ * With a block given, for each index:
+ *
+ * - If the index is in range, uses an element of +self+ (as above).
+ * - Otherwise, calls the block with the index and uses the block's return value.
+ *
+ * Example:
+ *
+ *   a = [:foo, :bar, :baz]
+ *   a.fetch_values(1, 0, 42, 777) { |index| index.to_s }
+ *   # => [:bar, :foo, "42", "777"]
+ *
+ * Related: see {Methods for Fetching}[rdoc-ref:Array@Methods+for+Fetching].
+ */
+
+static VALUE
+rb_ary_fetch_values(int argc, VALUE *argv, VALUE ary)
+{
+    bool block_given = rb_block_given_p();
+
+    long i, olen = RARRAY_LEN(ary);
+    VALUE result = rb_ary_new_capa(argc);
+    for (i = 0; i < argc; ++i) {
+        append_fetch_values_single(result, ary, olen, argv[i], block_given);
+    }
+    return RB_GC_GUARD(result);
```

Comment:
`result` is the return value, which shouldn't be optimized away.

Reply:
Fixed.